### ISSUE TITLE
Defining different forward/backward pass behavior

### ISSUE THREAD

**Issue by rcnlee** [JuliaDiff/ForwardDiff.jl#258]
I am trying to port Eric Jang's Gumbel-Softmax code from here:
http://blog.evjang.com/2016/11/tutorial-categorical-variational.html

The part that I am having trouble with is line 27 of the Tensorflow Implementation:
``y = tf.stop_gradient(y_hard - y) + y``

I believe the desired behavior is to use ``y_hard`` on the forward pass but use ``y`` to compute gradients on the backwards pass.

How can I achieve this in ForwardDiff (and ReverseDiff too)?

Thanks!
